source/openvr/openvr.cpp

```diff
@@ -422,3 +456,20 @@ HOOK_EXPORT void *   VR_CALLTYPE VR_GetGenericInterface(const char *pchInterface
 	
 	return interface_instance;
 }
+
+HOOK_EXPORT void* VR_CALLTYPE VRCompositor()
+{
+	static std::set<void*> installed;
```

Comment:
There should be no need for a `set` here. I believe this is only initialized on first call and then just returned, so can simply do:
```c++
HOOK_EXPORT vr::IVRCompositor *VR_CALLTYPE VRCompositor()
{
	static vr::IVRCompositor *compositor_instance = nullptr;
    if (compositor_instance == nullptr)
    {
        compositor_instance = reshade::hooks::call(VRCompositor)();
		// This is just for ProjectCARS2 / AMS2 and done through trial and error (IVRCompositor_010)
		// Need a way to work out what the Compositor version is. VR_GetGenericInterface isn't called so don't get to hook it there.
		reshade::hooks::install("IVRCompositor::Submit", vtable_from_instance(static_cast<vr::IVRCompositor*>(compositor_instance)), 4, reinterpret_cast<reshade::hook::address>(IVRCompositor_Submit_009));
    }
    return compositor_instance;
}
```

Reply:
Agreed, I should have updated this as it was to check if there were multiple calls.

In the absence of a way to get the interface versions such as with PC2 / AMS2, would it be worth adding an optional section to an ini file that allows OpenVR interface versions to be specified and then the hook function choice overriden if that section exists?

Reply:
I dug through the DLL a bit, and it appears `VRCompositor` is calling `IVRClientCore::GetGenericInterface` behind the scenes. The same function `VR_GetGenericInterface` also calls (https://github.com/ValveSoftware/openvr/blob/4c85abcb7f7f1f02adaf3812018c99fc593bc341/src/openvr_api_public.cpp#L185). So I think only thing we need to do is instead of hooking, `VR_GetGenericInterface`, hook `IVRClientCore::GetGenericInterface` and all problems would be solved.
Only issue is, how to hook the `IVRClientCore` interface ... Only way to do that I can see is to hook `VRClientCoreFactory` function from `vrclient.dll`/`vrclient_x64.dll` (which is loaded by `openvr_api.dll` on init). Will need to try that.

Anyway, for now this is a temporary solution that is fine. In future the `VR_Init(...)` + `VR_GetGenericInterface` hooks can be replaced with `VRClientCoreFactory` + `IVRClientCore::GetGenericInterface` hooks that should hopefully solve all versioning issues. Just need to find a game to test this with.